multiuser/api/che-multiuser-api-authentication-commons/src/main/java/org/eclipse/che/multiuser/api/authentication/commons/filter/MultiUserEnvironmentInitializationFilter.java

```diff
@@ -146,21 +157,35 @@ public void doFilter(ServletRequest request, ServletResponse response, FilterCha
   }
 
   /**
-   * Calculates id of the user from given authentication token. This <b>may</b> also imply
-   * verification of token signature or encryption for encrypted/signed tokens (like JWT etc)
+   * Processes the token and creates implementation-specific intermediary type using which the
+   * subclasses can extract different kinds of information like user ID or subject.
+   *
+   * <p>This <b>may</b> also imply verification of token signature or encryption for
+   * encrypted/signed tokens (like JWT etc).
+   *
+   * @param token the token to process
+   * @return a processed token or null if the token could not be processed for valid reasons (like
+   *     expiry or not matching any user).
+   */
+  protected abstract Optional<T> processToken(String token);
+
+  /**
+   * Retrieves the id of the user from given authentication token.
    *
-   * @param token authentication token
+   * @param token the original authentication token string
+   * @param processedToken the processed authentication string
    * @return user id given token belongs to
    */
-  protected abstract String getUserId(String token);
+  protected abstract String getUserId(String token, T processedToken);
 
   /**
    * Calculates user {@link Subject} from given authentication token.
    *
-   * @param token authentication token
+   * @param token the original authentication token string
+   * @param processedToken the processed authentication string
    * @return constructed subject
    */
-  protected abstract Subject extractSubject(String token) throws ServletException;
+  protected abstract Subject extractSubject(String token, T processedToken) throws ServletException;
```

Comment:
If I am not mistaken, after removeing token argument from `getUserId`, it can be reduced to `extractSubject(T processedToken)`

Reply:
The `token` is used in `extractSubject` for example here: https://github.com/eclipse-che/che-server/pull/136/files#diff-64f253ca97cefd0893104e2716d6a6d4452e4676bc99d6a45344e638a9b425afR100